### DISCUSSION TITLE
Brightness/Contrast adjustments with slider

### DISCUSSION THREAD

**Discussion by 4OTbIPE** [SixLabors/ImageSharp#1759] [Q&A]
Can you help me pls... How to implement the brightness change via the slider? Opening a file via a stream, changing its brightness via Mutate, opening a MemoryStream and saving an image to it with subsequent output to Image control is sooo slow. How to implement it correctly?
Its UWP
```
private async void BrightnessSlider_ValueChangedAsync(object sender, RangeBaseValueChangedEventArgs e)
        {
            if (CurrentImage != null)
            {
                using (Stream stream = await CurrentImage.File.OpenStreamForReadAsync())
                {
                    using (MemoryStream outStream = new MemoryStream())
                    {
                        using (SixLabors.ImageSharp.Image image = SixLabors.ImageSharp.Image.Load(stream))
                        {
                            image.Mutate(x => x.Brightness((float)BrightnessSlider.Value)
                                               .Contrast((float)ContrastSlider.Value));
                            await image.SaveAsBmpAsync(outStream);

                        }
                        byte[] arr = outStream.ToArray();
                        using (InMemoryRandomAccessStream str = new InMemoryRandomAccessStream())
                        {
                            using (DataWriter writer = new DataWriter(str.GetOutputStreamAt(0)))
                            {
                                writer.WriteBytes(arr);
                                await writer.StoreAsync();
                            }
                            BitmapImage bitmapImage = new BitmapImage();
                            await bitmapImage.SetSourceAsync(str);
                            CorrContent.Source = bitmapImage;
                        }
                    }
                }
            }
```

**Comment by JimBobSquarePants** [comment]
Without any code from you there’s no way to tell what you are doing or how to help you

**Reply by 4OTbIPE** [reply]
> Without any code from you there’s no way to tell what you are doing or how to help you

Oh, sorry. Updated

**Comment by antonfirsov** [comment]
This code does a couple of unnecessary operations on every value change:
- Decode the original image from a file (can be cached in memory)
- Encode the result image to a memory stream, and copy stream contents to a new `byte[]`
- Decode the `byte[]` with `SetSourceAsync()`

I'm sure this can be all avoided, but I'm not familiar enough with WPF to give specific advice how to do it. I can recommend this project implementing a WPF ImageSource on top of ImageSharp:
https://github.com/jongleur1983/SharpImageSource

Ideally, it should be even possible to do a zero-copy interop with `Image.WrapMemory<TPixel>(...)`, but that definitely needs some low-level WPF knowledge.

**Comment by AlexMauch** [comment]
Dear @antonfirsov,
 
I tried this and it did work well. But I am not that deep into ImageSharp or WPF, so I do not know wether there may be problems with this approach. I compared this with the mentioned "SharpImageSource" with a simple test and it seems to be at about 3 times faster. 

Would like to get a comment on wether I am heading into the right direction here.

```cs          
		[TestMethod] public void BitmapSource_WrapMemory()
		{
			Int32 width = 100;
			Int32 height = 100;
			Int32 bits = 32;
			Int32 stride = (width * bits + 7) / 8;
			Int32 bytes = stride * height;
			Byte[]? pixels = new Byte[bytes];

			var source = BitmapSource.Create(
				pixelWidth: width, 
				pixelHeight: height, 
				dpiX:  96, 
				dpiY: 96, 
				pixelFormat: PixelFormats.Bgra32, 
				palette: null, 
				pixels: pixels, 
				stride:  stride);

			var image = Image.WrapMemory<Argb32>(
				byteMemory: pixels, 
				width: width, 
				height: height);

			GraphicsOptions? options = GetOptions();

			image.Mutate(x => x.Fill(options, Color.AliceBlue));

			String? path = Helper.GetPath("BitmapSource_WrapMemory", "bmp");

			image.SaveAsBmp(path);
		}
```
